### PR TITLE
Bump NodeJS supported versions to ^16.16 || ^18.13 and add engines to package.json

### PR DESCRIPTION
**Pull request by alex-page**
### WHY are these changes introduced?

NodeJS 14 is EOL and we are removing support in v11.

We currently set `engines` on the root `package.json`. The problem with this is that consumers of the library do not get this information and they can use the libraries without any warning of what version of NodeJS it supports.

### WHAT is this pull request doing?

- [x] Increasing the supported versions of NodeJS
- [x] Add the engines field for NodeJS to package.json files

I would like to get NodeJS to version 16.19 before we launch version 11. However there is an issue with babel/rollup/browserslist not finding the latest version of NodeJS:
```
(plugin babel) BrowserslistError: Unknown version 16.19 of Node.js
``` 

### How to 🎩

- CI completes successfully

### CODE REVIEW

File: polaris-react/rollup.config.mjs [hunk 1]
```diff
@@ -54,7 +54,7 @@ function generateConfig({output, targets, stylesConfig}) {
 /** @type {import('rollup').RollupOptions} */
 export default [
   generateConfig({
-    targets: 'extends @shopify/browserslist-config, node 12.20',
```
Comment (alex-page) [inline polaris-react/rollup.config.mjs:57]: This scared me but nothing seems to be bad since changing the version...

File: .eslintrc.js [hunk 2]
```diff
@@ -116,12 +116,6 @@ module.exports = {
         'import/extensions': 'off',
         'import/no-default-export': 'off',
         'import/no-anonymous-default-export': 'off',
-        // We could omit this if we set `engines` fields properly
-        // As we don't set them then eslint thinks we're using node 8
-        'node/no-unsupported-features/node-builtins': [
-          'error',
-          {version: '>=16.0.0'},
-        ],
```
Comment (alex-page) [inline .eslintrc.js:124]: We set the engines field now so this is no longer needed.

**Review by sam-b-rose — APPROVED** [review]
Looks good! I did see that the `@types/node` in the `polaris-for-vscode/package.json` is on `^14`. I think it might be good to update that as well to `16` or even `18`.

File: documentation/guides/migrating-from-v10-to-v11.md [hunk 3]
```diff
@@ -4,11 +4,16 @@ Polaris v11.0.0 ([full release notes](https://github.com/Shopify/polaris/release
...
 
+- [Node support](#node-support)
 - [Components](#components)
   - [Removed deprecated Collapsible argument](#removed-deprecated-collapsible-argument)
   - [Removed KonamiCode component](#removed-konamicode-component)
 - [Tokens](#tokens)
 
+## Node support
+
+NodeJS version 14 is no longer supported. NodeJS 18 is recommended and 16 is supported.
```
Comment (alex-page) [inline documentation/guides/migrating-from-v10-to-v11.md:15]:
```suggestion
NodeJS version 14 is no longer supported. NodeJS 18 is recommended and 16 is the minimum supported version.
```

File: polaris-for-vscode/package.json [hunk 4]
```diff
@@ -20,7 +20,7 @@
     "shopify"
   ],
   "engines": {
-    "vscode": "^1.64.0"
+    "node": "^16.16.0 || ^18.13.0"
```
Comment (BPScott) [inline polaris-for-vscode/package.json:23]: This seems incorrect

File: polaris-cli/package.json [hunk 5]
```diff
@@ -31,7 +31,7 @@
   },
   "engine-strict": true,
   "engines": {
-    "node": ">=14.13.1"
+    "node": "^16.16.0 || ^18.13.0"
```
Comment (BPScott) [inline polaris-cli/package.json:34]: By setting the engines on the specific node packages, they shall refuse to install if the node version does not match these versions. This is great for blocking installation on older unsupported versions, however the configuration of "^16.16.0 || ^18.13.0" means "Only node 16 or 18" - you're currently also blocking the installation of these packages on  the currently supported node 19 (and the lts node 20 when it comes out in April).

For these per-package engine fields I strongly think we should allow current and future versions for the sake of forwards compatibility - `^16.16.0 || >=18.13.0`.

**Review by BPScott — COMMENTED** [review]
The versions chosen seem suprising to me

- Using "^16.16.0 || ^18.13.0" for the per-package engines shall block installation on node 19 (and 20 when it comes out in April). Currently supported and future node versions seems like a bad idea.
- Node 14 is still supported [till April](https://github.com/nodejs/Release). I'm ok with preemptively dropping support if you feel like this shall help improve something.
- The first LTS version of v16 was [16.13](https://github.com/nodejs/node/blob/main/doc/changelogs/CHANGELOG_V16.md), and the first LTS version of 18 was [18.12](https://github.com/nodejs/node/blob/main/doc/changelogs/CHANGELOG_V18.md). Is there a reason why we're choosing to claim minimum polaris versions are 16.16 and 18.13?

**Pull request merged**